### PR TITLE
LockFile Feature

### PR DESCRIPTION
**Pull request by kodypeterson**
Based on https://github.com/bower/bower/pull/1592

---
- bower.lock doesn't exist:
  - [x] bower install creates bower.lock file only after all-successful non-production installation
  - [x] bower install --production complains there's no bower.lock and exits
  - [x] bower update acts as bower install
- bower.lock exists dependenies in bower.json match those on bower.lock
  - [x] bower install installs exact packages from bower.lock
  - [x] bower install --production installs exact packages from bower.lock ignoring devDependencies
  - [x] bower update is no-op
  - [x] if installation would change bower.lock, it fails with error message
- bower.lock exists and there are only extra dependencies in bower.json relative to bower.lock
  - [x] bower install tries to install bower.lock with only new packages from bower.json added. Installation fails if old bower.lock is not a subset of new bower.lock.
  - [x] bower install --production fails
  - [x] bower update --all re-creates bower.lock
- bower.lock` exists and dependencies in bower.json changed relative to bower.lock
  - [x] bower install fails, and says you need to bower update <package> first
  - [x] bower install --production fails
  - [x] bower update complains there's no package specified to update
  - [x] bower update --all re-creates bower.lock
  - [x] bower update <package> tries to install bower.lock with <package> version replaced with version from bower.json. Installation fails if old bower.lock is not a subset of new bower.lock (except updated package).

### CODE REVIEW

**Comment by kodypeterson** [comment]
@sheerun - So, when running `bower update` and nothing has yet to be installed, should a lock file be required as at that point it really acts like `bower install`?

**Comment by kodypeterson** [comment]
@sheerun bump for the question above... :smile:

**Comment by sheerun** [comment]
@kodypeterson Sorry for the delay. Yes, indeed it should do the same as `bower install` :)

**Comment by kodypeterson** [comment]
@sheerun any thoughts on the lock file being .bower.lock instead of bower.lock?

Just wondering if we really want to make it a non-dot file...

**Comment by kodypeterson** [comment]
Also, I noticed that most of the update tests are invalid as they call tempDir.prepare({JSON}) after the install. Which .prepare will remove all of the contents in the directory.

So, when update runs, it is running against an empty directory with no dependencies installed, so it just acts like `bower install` :frowning:

**Comment by kodypeterson** [comment]
How should bower link be handled, it in turn calls bower update? Should it follow the same rules as bower update?

**Comment by sheerun** [comment]
`bower link` is totally separate thing. It just creates a symlink to some other directory.

I think we don't need to worry about it.

**Comment by kodypeterson** [comment]
@sheerun  Are you sure? :smile: 

https://github.com/bower/bower/blob/master/lib/commands/link.js#L62

**Comment by sheerun** [comment]
Huh. I'm not sure if it's necessary. Maybe we should check what NPM does..

**Comment by kodypeterson** [comment]
@sheerun - It seems it runs an install https://github.com/npm/npm/blob/master/lib/link.js#L134

For now, I have left it as is. Allowing it to run update and allowing the update to run without allowing it to make changes to the lockfile. Your call on if that is correct.

**Comment by sheerun** [comment]
I think most reasonable is to just call install, instead of of update.
Install doesn't change lockfile by default.

On Tue, Mar 31, 2015 at 11:20 AM, Kody J. Peterson <notifications@github.com

**Comment by kodypeterson** [comment]
Alright, I think I have everything covered here :smile: 

Time for some regression testing? :smile: 

Not sure why there is a -0.12% decrease in code coverage.... Do I need to add more tests or is this going to be ok?

**Comment by sheerun** [comment]
Yeah, I think we need some more tests and peer review on this. I'll try to review it soon.

Could you squash the commits in meaningful pieces?

What about "if installation would change bower.lock, it fails with error message"? This one is actually pretty important as it ensures that production builds are reproducible. For example it should fail when someone changes the commit behind a tag on GitHub.

**Comment by kodypeterson** [comment]
Ah! That is what that means. I will add that functionality, no problem!

I will also add more tests somehow..... Most of the lockfile logic is tested... I will see what else I can add.

**Comment by satazor** [comment]
@kodypeterson why not use a lock module instead? See: https://github.com/IndigoUnited/node-proper-lockfile and https://www.npmjs.com/package/lockfile

**Comment by kodypeterson** [comment]
@satazor At first glance it seems that module is for locking a specific file... That is not really what we are trying to accomplish... Am I wrong in what that module does?

**Comment by satazor** [comment]
Both modules work by locking a file yes, but isn't what your doing? You are creating a bower.lock to lock down a project. Your lock file strategy is using .writeFile which isn't atomic. What I'm saying is the file locking is not an easy matter and it should be delegated to more mature modules such as those two that I mentioned.

**Comment by kodypeterson** [comment]
@satazor We are locking down versions of libraries you are using. We want to be sure that when installing that you are always installing the same versions of dependencies and such that you did when you were developing locally. We are creating a very similar feature to NPM's shrinkwrap.

**Comment by satazor** [comment]
Oh I totally misunderstood this then. Just ignore me.

File: lib/core/Project.js [hunk 1]
```diff
@@ -135,84 +141,88 @@ Project.prototype.update = function (names, options) {
...
-            return that._manager.install(that._json);
-        })
-        .then(function (installed) {
-            // Save JSON, might contain changes to resolutions
-            return that.saveJson()
-            .then(function () {
-                return that._manager.postinstall(that._json).then(function () {
-                    return installed;
-                });
+                //console.log(targets[0]);
```
Comment (destroyerofbuilds) [inline lib/core/Project.js]: Can this console statement be removed?

**Comment by kodypeterson** [comment]
@sheerun - I am working on adding that last feature and a few more tests. Also removing all console logs!!! I have found that I am really good at leaving those in. Hopefully we should be ready to really test this guy out and get it shipped! :smile:

**Comment by thefallentree** [comment]
:+1:

**Comment by foglerek** [comment]
This is great, thanks for all your hard work on this. Looking forward to seeing it implemented.

**Comment by kodypeterson** [comment]
@sheerun - I think we are all good here! :smile: 2 months later ;)

**Comment by kodypeterson** [comment]
@sheerun - bump ^^

**Comment by sheerun** [comment]
I'm pretty busy these days, but I'll try to review soon

**Comment by sheerun** [comment]
Or maybe someone else can? :)

**Comment by kitak** [comment]
:+1:

**Comment by fourseven** [comment]
I would like to point out the discussion on the bundler github - https://github.com/bundler/bundler/issues/694 - that they're moving away from the .lock suffix and it's probably worth considering following suit, as it's always clashed with old *NIX conventions.

**Comment by aequasi** [comment]
Any reason this hasnt been merged yet?

**Comment by kodypeterson** [comment]
@aequasi No one has had time to review it :disappointed: @sheerun anyone you can get to review this?

**Comment by sheerun** [comment]
Not at least until 1st July.. Sorry.

**Comment by thefallentree** [comment]
ping? already July13!

**Comment by sheerun** [comment]
Currently I'm working on deploying pluggable resolvers feature. If someone has time to review this, please do it.

**Comment by sheerun** [comment]
sheerun@sher.pl sent you an invitation to join MEGA.\n\n";

--- Message from user: ---
Hello, join me on MEGA and get access to encrypted storage and
communication. Get 50 GB free!
---

Please create your free MEGA account by clicking on the following
link:\n\n";
https://mega.nz/#newsignupcmVwbHkrMDAwNDc2MGQxNDkwYmU5YmM0MmFmNThiZmNjNmU1MDA2MzljNjVjZTBkYjU2NTBhOTJjZjAwMDAwMDAxMTEzMmE3ODE5MmExNjljZTAzYzc4Mjc4QHJlcGx5LmdpdGh1Yi5jb2236xFCh2nnIA

Best regards,

Team MEGA

**Comment by aequasi** [comment]
@sheerun hacked?

**Comment by johnpapa** [comment]
+1 on this

This is holding us up too and we have a hack to get around it.

**Comment by tuwannu** [comment]
I really wish to see this feature released. Is there something for a non-contributor so far to help on this?

**Comment by sheerun** [comment]
@tuwannu You could checkout this branch locally, test different scenarios, and report issues with it :) There must be some.

**Comment by enumag** [comment]
It would be better for testing if the auther rebased it first.

**Comment by kodypeterson** [comment]
@enumag @sheerun I will get it rebased for testing. Can we get a release scheduled for this please! :smile:

**Comment by tuwannu** [comment]
@kodypeterson I'm standing-by to help testing after the rebase. Also made a pull request a few days ago to your repo to fix a couple of unit tests.

**Comment by syzer** [comment]
:+1:

**Comment by lsmith77** [comment]
not sure if I understood your description properly but I think you could simplify the logic for detecting mismatches between the bower.json and bower.lock file by simply storing a hash of the bower.json file in your bower.lock file. that way you can always warn the user if someone has changed the bower.json file since the bower.lock file was generated. afaik this is how the composer package manager works for PHP.

**Comment by twogee** [comment]
+1

**Comment by jpike88** [comment]
For the love of god approve this! This will make bower kick ass!

**Comment by fesor** [comment]
:+1:

**Comment by mkusher** [comment]
+1

**Comment by schmurfy** [comment]
can we please stop with the +1 ? Just subscribe if you want to know when/if anything happens.

**Comment by lsmith77** [comment]
since this is not getting any attention from the core devs it seems, one option would be to refactor this into a custom resolver. but honestly the default behavior of bower is so far away from deterministic behavior I wonder if its really worth it to try and salvage it.

**Comment by jpike88** [comment]
I'm actually kind of blown away that this mechanism wasn't included in bower from the beginning. It's such a fundamental aspect of externally sourced dependency management, you wonder how bower got this far.

**Comment by syzer** [comment]
most-evil-bower

**Comment by greg0ire** [comment]
lol people are writing [blog posts](https://blog.liip.ch/) about this. Has this issue the highest priority ? Because it looks like it should… I think many people are leaving Bower just because of this.

**Comment by Rvanlaak** [comment]
@syzer you can now also admire your illustration in the blog post https://blog.liip.ch/archive/2015/10/16/bower-kthxbye.html

Lockfile would give bower the boost it needs right now.

**Comment by vincentwoo** [comment]
@sheerun, can you comment as to whether this PR is under active review? A status update on your plans here would be very appreciated. I know a lot of people affected by this issue and there hasn't been much in the way of official communication lately.

**Comment by sheerun** [comment]
Currently it's not under review by me. There are currently no other active core members in Bower.

I'm very interested in this feature as well, but it would require tenths of hours for review and finish. And then even more for maintenance after release. It probably means that Bower needs to seek stable financial support first, so it can provide appropriate support. Even when people that are currently pushing for it, become uninterested in maintaining it (possible, and even very likely to happen).

The main reason Bower got nice Pluggable Resolvers is because Artifactory agreed to sponsor it for this feature. Without it we'd need to accept [original PR for them](https://github.com/bower/bower/pull/1686), created with minimal effort, and then get far worse API for and maintenance issues afterwards. The same applies to this feature. To release it we need to be sure that it is stable, future-proof, and there's someone to maintain it.

As a first step I'll try to find some sponsors, so https://salt.bountysource.com/teams/bower no longer says "$0 this month from 0 supporters". The next (easier) step is to find developers ready to develop and maintain this feature for little money we get. Personally I find it hilarious how much Bower is used and how little support it gets. Community comments about "killing it as soon as possible" don't help as well.

**Comment by vincentwoo** [comment]
@sheerun, your update is well taken.

A few points of unsolicited feedback, from me as just some random guy on the internet (who has helped fund a few OSS projects) to you, the apparent sole maintainer of a very popular open source project:

* The structure of the "organization" behind bower is completely opaque from the website. When I started using bower, I clicked around bower.io, noted that it seemed professionally (if minimally) produced, and that there was a lot of community support for it. I assumed there was a team and probably at least one serious corporate sponsor. The about page says "Bower is a team effort," which is a statement that I now regard as generally untrue.
* That there *no maintainers* besides you, who cannot afford to spend a lot of time on the project is a *serious problem*, and one that is not made apparent except through by explicitly pestering you in this comment thread. I would suggest making a very prominent call to action to get either the funds or support you need **right on the homepage**. The bountysource page you linked to I cannot find anywhere on bower.io.
* Bower is clearly important enough to the infrastructure of enough major websites that you can probably flat out charge them for access to even *just* this feature or enterprise support services, etc.

**Comment by sheerun** [comment]
I've prepared a Google Form for anyone interested supporting Bower. Please fill it if you can:

http://goo.gl/forms/P1ndzCNoiG

**Comment by 304NotModified** [comment]
+1. 

Showstopper this.

**Comment by bigdot** [comment]
@sheerun can you give me a $number for you to round this feature up ?

**Comment by sheerun** [comment]
I think ~$5000 is a reasonable number (for implementing and then maintaining for some time). But I won't so that it as I don't have enough time. Maybe @kodypeterson @faceleg or @albertinad could tell?

**Comment by nnjpp** [comment]
@sheerun claims of bower's death are spreading really fast right now across HN, Reddit, and individual projects. I would suggest a brief PR statement that puts the myths to rest and includes a call to action for the funding this project needs, otherwise usage will probably drop due to FUD.

**Comment by AdamWills** [comment]
@nnjpp @sheerun Came here because of this.. from reddit via a tweet. Some clarification on the state of Bower would be great... and perhaps this publicity could help the project.

**Comment by wrumsby** [comment]
I hate adding +1 comments, but +1 @nnjpp @AdamWills.

**Comment by faceleg** [comment]
The bower project is seeking both funding and active contributors.

**Comment by sheerun** [comment]
If it's not clear yet: bower is stable software, bower is not dying, bower looks for contributors and funding.

**Comment by vincentwoo** [comment]
If you are seeking contributors and funding you should really, really make
that clear on the website instead of hoping people stumble onto this thread.

**Comment by sheerun** [comment]
@vincentwoo For now it's only in the README. At the very top.

**Comment by potherca** [comment]
I agree with @vincentwoo on this. I've opened [a PR for adding a callout banner to the website](https://github.com/bower/bower.github.io/pull/175).

**Comment by Utsav2** [comment]
@kodypeterson Great work on this. 

As @sheerun mentioned before, is it possible to squash this into meaningful commits and rebase against master? It's hard to review in its current state.

**Comment by sheerun** [comment]
@Utsav2 Could you do it instead and send another PR?

**Comment by Utsav2** [comment]
#2100

**Comment by iBasit** [comment]
No project or version of any project is ever completed without any bugs. Thats why versions keep bumping up for any repository.

You should release as dev version and contributor would be happy to fix and maintain this feature. 

For more then 3+ years, this feature is been pending and 1 year for review as contributor completed his task. 

**Please just release this under dev version and let everyone else review it, test it and give there feedback.**

**Comment by iBasit** [comment]
If you need any help with testing, please feel free to message me.

**Comment by shyiko** [comment]
For what it's worth shrinkwrap functionality is available using [bower-shrinkwrap-resolver](https://github.com/shyiko/bower-shrinkwrap-resolver). You might find it useful until this PR gets merged in.

**Comment by nicodmf** [comment]
+1

**Comment by nicodmf** [comment]
In my opinion, if bower lock exist bower install shouldn't fails, but just install the bower lock content with information message.

**Comment by kodypeterson** [comment]
Alright @sheerun what are we going to do with this one? We missed the birthday celebration of this PR... Soon, it will be 2 years old.

It now has a bunch of conflicts, which I can work to resolve. Are the requirements above still accurate? If so, I can work to get a new PR started. If not, can we get a new issue opened with the correct requirements laid out so I can get to work on it. We need to get this functionality into bower.

It is very obvious the demand for it, and I am here to make it happen! Just let me know what I can do to get this going again!

PS: Sorry for dropping the ball on this PR, it sort of fell of my radar.

**Comment by sheerun** [comment]
@kodypeterson To release it as quickly as possible we'd need to put it in current 1.x version of bower, and make it backward compatible (opt-in). What sounds reasonable is:

- Squash all commits into one (it makes rebase easier)
- Rebase this PR onto master
- Add `--lock` flag to `bower install` to enable this feature when bower.json already exists
- Add a warning if there isn't a lockfile available, prompting to use `--lock` flag
- After bower.lock has been created, keep implemented behavior
- Do not modify any existing tests and only create new ones (when `bower.lock` is present)
- Make sure all current tests pass

If we ever release 2.0, we set `--lock` to be a default.

As for lock itself, I see it contains few extra keys we don't want, especially they contain absolute paths:

- `endpoint`, `canonicalDir`

```
  "endpoint": {
    "name": "bower",
    "source": "/Users/sheerun/Source/Bower/bower",
    "target": "1.7.9"
  },
  "canonicalDir": "/Users/sheerun/Source/Bower/bower",
```

Also, we need to put version of bower it was packaged with to `version` key, plus additional behavior:

- If bower minor version in bower.lock is greather than current bower version, refuse to install package, indicating that newer bower version is necessary.

I've also encountered an error when installing local package without bower.json:

```
Stack trace:
TypeError: Cannot read property 'pkgMeta' of undefined
    at /Users/sheerun/Source/Bower/bower/lib/util/lockFile.js:22:84
    at /Users/sheerun/Source/Bower/bower/node_modules/mout/object/forOwn.js:12:27
```

I can do more testing if we you manage to fix those. Thank you for being interested in continuing your work.

**Comment by sheerun** [comment]
@kodypeterson Also, then I did "bower install angular-route", only "angular-route" got saved to bower.lock, while it should also contain "angular".

**Comment by kodypeterson** [comment]
@sheerun I am on it, I will see what I can get through this weekend along with cleanup and such!

**Comment by Utsav2** [comment]
@kodypeterson I fixed it up into one commit and tried rebasing it some time
back in #2100, it might be useful to get started there.

**Comment by kodypeterson** [comment]
@Utsav2 awesome! i will see what I can get from that. I have a feeling that both being so long ago the merge process is going to be a rough one. I might just need to cherry pick manually the changes in and maybe reorganize a bit. 

Thanks for the reference though!

**Comment by filib** [comment]
@kodypeterson need any help?

**Comment by sheerun** [comment]
Maybe a better route is to fix bower support in https://yarnpkg.com/ that already has file locking

**Comment by filib** [comment]
@sheerun I'm not entirely sure how practical that is; we're using PureScript which leverages Bower for dependency management and it looks like support for Bower in Yarn is really just in its infancy e.g. https://github.com/yarnpkg/yarn/pull/896.

Does this mean that you would no longer be willing to consider this patch?

**Comment by sheerun** [comment]
Maybe we could at least adopt lockfile format of yarn?

**Comment by joelhandwell** [comment]
For those who want to use lock feature before this PR merged, [bower-locker](https://github.com/infusionsoft/bower-locker) is a workaround to generate lockfile and commit only ```bower.json``` which is newly generated lock file and ```bower-locker.bower.json``` which is renamed original ```bower.json``` in your repo.


Run following commands to achieve it:

    npm install bower-locker -g

or 

    yarn global add bower-locker

then generate lock file based on existing ```bower.json``` file by runing: 

    bower-locker lock

**Comment by kael-shipman** [comment]
Any news on this? Is it stalled out?

**Comment by twogee** [comment]
Why is this PR open when there is #2100?

**Comment by sheerun** [comment]
Bower is depreacted now and we recommend switching to Yarn that supports version locking and more. Here's guide how to migrate: https://bower.io/blog/2017/how-to-migrate-away-from-bower/

**Pull request closed without merging**